File: FindKey.java
package CS211;

import java.math.*;
/**
 * This is a program to find the key and the message of a ElGamal encryption by brute force.
 * @author Corentin Pinato (14739811)
 * @version 1.0
 */
public class FindKey 
{
	public static void main(String args[])
	{
    	BigInteger publicKey[] = {new BigInteger("24852977"),new BigInteger("2744"),new BigInteger("8414508")};
	    BigInteger message[] = {new BigInteger("15268076"),new BigInteger("743675")};
		
		BigInteger Message = new BigInteger("1");

		BigInteger encrypted[] = encrypt(Message,publicKey,new BigInteger("1"));
	
		BigInteger key = new BigInteger("1");
		
		BigInteger decrypted = decrypt(publicKey,encrypted,key);
		while(decrypted.compareTo(Message) != 0)
		{
			key = key.add(new BigInteger("1"));
			decrypted = decrypt(publicKey,encrypted,key);
		}	
		System.out.println("The Private Key is: "+key);
		System.out.println("So the secret message is: "+decrypt(publicKey,message,key));
	}
	public static BigInteger decrypt(BigInteger publicKey[], BigInteger message[], BigInteger privateKey)
	{
		BigInteger p = publicKey[0];
		
		BigInteger gy = message[0];
		BigInteger mgxy = message[1];
		
		BigInteger p1x = p.subtract(new BigInteger("1")).subtract(privateKey);
		
		BigInteger c1 = gy.modPow(p1x, p);
		
        BigInteger c2 = mgxy.multiply(c1);
		c2 = c2.mod(p);
		
		return c2;
	}
	public static BigInteger[] encrypt(BigInteger message, BigInteger publicKey[], BigInteger randomNum)
	{
		BigInteger p = publicKey[0];
		BigInteger g = publicKey[1];
		BigInteger gxmodp = publicKey[2];
		

		BigInteger m1 = g.modPow(randomNum, p);
		BigInteger m2 = message.multiply(pow(gxmodp,randomNum)).mod(p);
		
		BigInteger result[] = {m1,m2};
		
		return result;
	}
	public static BigInteger pow(BigInteger num, BigInteger power)
	{
		BigInteger result = num;

		while( power.compareTo(new BigInteger("1"))>0)
		{
			result = result.multiply(num);
			power = power.subtract(new BigInteger("1"));
		}
		return result;
	}
}

File: FindKey.py
# Corentin Pinato
# This is program has to find the decryption key of an ElGamal encryption method by using a brute force strategy.

#publicKey = [8347852664524685394671539,1725419847553,6544719840647906985061541]
#message = [2030735454253481748649532,1340127105753313239218903]

publicKey = [29, 2, 3]
message = [23, 27]

#publicKey = [24852977,2744,8414508]
#message = [15268076,743675]

def decrypt(publicKey, message, privateKey):
    p = publicKey[0]
    g = publicKey[1]
    gxmodp = publicKey[2]

    gy = message[0]
    mgxy = message[1]

    c1 = (gy**(p-1-privateKey))%p
    c2 = (mgxy*c1)%p

    return c2

def encrypt(message,publicKey,randomNum):
    p = publicKey[0]
    g = publicKey[1]
    gxmodp = publicKey[2]

    m1 = (g**randomNum)%p
    m2 = (message*(gxmodp**randomNum))%p

    result = [m1,m2]

    return result

def modPow(number,power,mod):

    if(power == 0):
        return 1
    elif(power%2==0):
        halfpower=modPow(number,(power/2),mod)
        return modMult(halfpower,halfpower,mod)
    else:
        halfpower=modPow(number,((power)/2)+1,mod)
        firstbit = modMult(halfpower,halfpower,mod)
        return modMult(firstbit,number,mod)


def modMult(first,second,mod):
    if(second==0):
        return 0
    elif(second%2==0):
        half=modMult(first,(second/2)+1,mod)
        return(half+half)%mod
    else:
        half=modMult(first,((second)/2)+1,mod)
        return(half+half+first)%mod


Message = 1

encrypted = encrypt(Message,publicKey,2)
print(encrypted)

key = 1

decrypted = decrypt(publicKey,encrypted,key)
print("Tryed key = "+str(key))

while(decrypted != Message):
    key += 1
    decrypted = decrypt(publicKey,encrypted,key)
    print("Tryed key = "+str(key))

print("The key is: "+str(key))

print("So the secret Message is: "+str(decrypt(publicKey,message,key)))